oeps/oep-0006.rst

```diff
@@ -0,0 +1,224 @@
+
+
+Specification
+=============
+
+New BlockScope values
+---------------------
+
+To support context-scoped fields, ``BlockScope`` will allow these new additional values:
+
+* ``CONTEXT_TYPE``: The data is related to all instances of this type of XBlock within a particular context.
+* ``CONTEXT_ALL``: The data is accessible to XBlocks of any type, within a particular context. (As with ``BlockScope.ALL``, XBlock authors who use such fields should be careful to use a namespace that will not conflict with ``BlockScope.CONTEXT_ALL`` fields defined by other XBlock authors.)
+
+A "context" is defined by the runtime but is usually a "course" or the nearest equivalent. In the Open edX platform, the context may be a course or a content library. Different runs of the same course would be considered different contexts and would not share data, though UserScope.NONE fields would be copied when creating a new course run. On other platforms, a "context" could be defined as a blog post or a textbook.
+
+New named scopes
+----------------
+
+As is the situtation today, XBlock runtimes would generally not be expected to implement every possible combination of ``BlockScope`` and ``UserScope``. Instead, XBlock runtimes are encouraged to implement these new named scopes, which would be defined in the XBlock python package:
+
+* ``Scope.context_settings`` (UserScope.NONE, BlockScope.CONTEXT_TYPE)
+
+  * Example: For an LTI XBlock, a list of the LTI provider URLs and secrets used in the course
+  * Example: For a discussions XBlock: Enable or disable discussions in this course
+
+* ``Scope.context_shared_settings`` (UserScope.NONE, BlockScope.CONTEXT_ALL)
+
+  * Example: Course-specific defaults that can apply to many different blocks, such as "max_attempts" and "show_answer", which would be consumed by the Capa Problem XBlock, Drag and Drop problem XBlock, Problem Builder XBlock family, etc. (These are currently implemented via ``InheritanceMixin``, which is built-in to edx-platform and does not allow XBlock plugins to extend the list of inherited fields. However, a downside of making them a context-scoped field is that each block would require two field definitions: a ``Scope.settings`` field for the usage-specific value, and a ``Scope.context_shared_settings`` field for the default/fallback value. )
+  * Example: Submission blackout dates - don't allow users to submit answers during the specified time period.
+
+* ``Scope.context_preferences`` (UserScope.ONE, BlockScope.CONTEXT_TYPE)
+
+  * Example: Enable or disable notifications about new discussion posts in this course
+  * Example: Video playback speed for this course
+  * Example: An XBlock in which students "accumulate building blocks over the course of the semester" (`PLAT-325`_)
```

Comment:
Sure, the first example isn't great if there is a better mechanism for that.

> As for the video example -- I suppose we could make the case that there are some subjects or languages where a student might want that flexibility, but I think that's an edge case. 

I mentioned that from personal experience - I normally watch videos at 1x, but I took one course on edX where the prof (who was talking throughout most of the videos) spoke slowly and took long pauses between every clause that he spoke - so for that particular course I sped the videos up to make the frequent silences less annoying.

> I'm concerned that the most compelling use cases I can think of for this scope make it very easy for developers to shoot themselves in the foot.

Good points. Limiting the allowed data size (and/or type?) seems like a way of reducing the potential impact of these problems, though I do worry that developers might not be aware of the limit at first, and build and deploy an XBlock that keeps a growing log, only to find out about the length limit when it's in production.